File: docs-source/docs/modules/microservices-tutorial/examples/02-shopping-cart-service-java/src/main/java/shopping/cart/ItemPopularityProjectionHandler.java
package shopping.cart;

import akka.Done;
import akka.actor.typed.ActorSystem;
import akka.projection.eventsourced.EventEnvelope;
import akka.projection.javadsl.Handler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class ItemPopularityProjectionHandler extends Handler<EventEnvelope<ShoppingCart.Event>> { // <1>
    private final Logger logger = LoggerFactory.getLogger(getClass());
    private final String tag;
    private final ItemPopularityRepository repo;

    public ItemPopularityProjectionHandler(String tag, ItemPopularityRepository repo) {
        this.tag = tag;
        this.repo = repo;
    }

    @Override
    public CompletionStage<Done> process(EventEnvelope<ShoppingCart.Event> envelope) throws Exception, Exception { // <2>
        ShoppingCart.Event event = envelope.event();

        CompletionStage<Done> dbEffect = null;
        if (event instanceof ShoppingCart.ItemAdded) {
            ShoppingCart.ItemAdded added = (ShoppingCart.ItemAdded) event;
            dbEffect = this.repo.update(added.itemId, added.quantity);
        } else if (event instanceof ShoppingCart.ItemQuantityAdjusted) {
            ShoppingCart.ItemQuantityAdjusted adjusted =
                    (ShoppingCart.ItemQuantityAdjusted) event;
            dbEffect = this.repo.update(adjusted.itemId, adjusted.newQuantity - adjusted.oldQuantity);
        } else if (event instanceof ShoppingCart.ItemRemoved) {
            ShoppingCart.ItemRemoved removed = (ShoppingCart.ItemRemoved) event;
            dbEffect = this.repo.update(removed.itemId, -removed.oldQuantity);
        } else {
            // skip all other events, such as `CheckedOut`
            dbEffect = CompletableFuture.completedFuture(Done.getInstance());
        }

        dbEffect.thenAccept(done -> logItemCount(event));

        return dbEffect;
    }

    /** Log the popularity of the item in every `ItemEvent` every `LogInterval`. */
    private void logItemCount(ShoppingCart.Event event) {
        if (event instanceof ShoppingCart.ItemEvent) {
            ShoppingCart.ItemEvent itemEvent = (ShoppingCart.ItemEvent) event;

            String itemId = itemEvent.itemId;
            repo.getItem(itemId)
                .thenAccept(
                    opt ->
                        logger.info(
                                "ItemPopularityProjectionHandler({}) item popularity for '{}': [{}]",
                                this.tag,
                                itemId,
                                opt.orElse(0L))
                );
        }
    }
}

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityProjectionHandler.scala
// tag::handler[]
package shopping.cart

import akka.actor.typed.ActorSystem
import akka.projection.eventsourced.EventEnvelope
import akka.projection.jdbc.scaladsl.JdbcHandler
import org.slf4j.LoggerFactory
import shopping.cart.repository.ScalikeJdbcSession

class ItemPopularityProjectionHandler(
    tag: String,
    system: ActorSystem[_],
    repo: ItemPopularityRepository)
    extends JdbcHandler[
      EventEnvelope[ShoppingCart.Event],
      ScalikeJdbcSession]() { // <1>

  private val log = LoggerFactory.getLogger(getClass)

  override def process(
      session: ScalikeJdbcSession,
      envelope: EventEnvelope[ShoppingCart.Event]): Unit = { // <2>
    envelope.event match { // <3>
      case ShoppingCart.ItemAdded(_, itemId, quantity) =>
        repo.update(itemId, quantity)
        logItemCount(itemId)

      // end::handler[]
      case ShoppingCart.ItemQuantityAdjusted(
            _,
            itemId,
            newQuantity,
            oldQuantity) =>
        repo.update(itemId, newQuantity - oldQuantity)
        logItemCount(itemId)

      case ShoppingCart.ItemRemoved(_, itemId, oldQuantity) =>
        repo.update(itemId, 0 - oldQuantity)
        logItemCount(itemId)

      // tag::handler[]

      case _: ShoppingCart.CheckedOut =>
    }
  }

  private def logItemCount(itemId: String): Unit = {
    log.info(
      "ItemPopularityProjectionHandler({}) item popularity for '{}': [{}]",
      tag,
      itemId,
      repo.getItem(itemId).getOrElse(0))
  }

}
// end::handler[]
